From 8be7f28707394d16c19d415c6506b67c318586de Mon Sep 17 00:00:00 2001
From: "Philip, Avinash" <avinashphilip@ti.com>
Date: Tue, 14 Aug 2012 13:48:55 +0530
Subject: [PATCH 5/6] PWM: ecap: Fix for throwing PWM output before running

On configuring duty cycle, APWM mode is enabled and left enabled even if
PWM not running. Fixes the same by disabling APWM mode if PWM device not
running. This ensures PWM output is low if PWM device not running.

Signed-off-by: Philip, Avinash <avinashphilip@ti.com>
---
 drivers/pwm/ecap.c | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/drivers/pwm/ecap.c b/drivers/pwm/ecap.c
index 670b32e..864f3ee 100644
--- a/drivers/pwm/ecap.c
+++ b/drivers/pwm/ecap.c
@@ -131,16 +131,13 @@ static int ecap_pwm_set_polarity(struct pwm_device *p, char pol)
 
 static int ecap_pwm_config_period(struct pwm_device *p)
 {
-	unsigned long flags, v;
+	unsigned long flags;
 	struct ecap_pwm *ep = to_ecap_pwm(p);
 
 	 pm_runtime_get_sync(ep->dev);
 
 	spin_lock_irqsave(&ep->lock, flags);
 	writel((p->period_ticks) - 1, ep->mmio_base + CAPTURE_3_REG);
-	v = readw(ep->mmio_base + CAPTURE_CTRL2_REG);
-	v |= (ECTRL2_MDSL_ECAP | ECTRL2_SYNCOSEL_MASK);
-	writew(v, ep->mmio_base + CAPTURE_CTRL2_REG);
 	spin_unlock_irqrestore(&ep->lock, flags);
 
 	pm_runtime_put_sync(ep->dev);
@@ -167,6 +164,12 @@ static int ecap_pwm_config_duty(struct pwm_device *p)
 	}
 	spin_unlock_irqrestore(&ep->lock, flags);
 
+	if (!pwm_is_running(p)) {
+		v = readw(ep->mmio_base + CAPTURE_CTRL2_REG);
+		v &= ~ECTRL2_MDSL_ECAP;
+		writew(v, ep->mmio_base + CAPTURE_CTRL2_REG);
+	}
+
 	pm_runtime_put_sync(ep->dev);
 	return 0;
 }
-- 
1.8.1.4

